fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoveAggregationFromAggTable.java

```diff
@@ -0,0 +1,120 @@
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class RemoveAggregationFromAggTable extends TransformationRule {
+    private static List<String> unsupportedFunctionNames =
+            ImmutableList.of(FunctionSet.BITMAP_UNION,
+                    FunctionSet.BITMAP_UNION_COUNT,
+                    FunctionSet.HLL_UNION,
+                    FunctionSet.HLL_UNION_AGG,
+                    FunctionSet.PERCENTILE_UNION);
+
+    public RemoveAggregationFromAggTable() {
+        super(RuleType.TF_REMOVE_AGGREGATION_BY_AGG_TABLE,
+                Pattern.create(OperatorType.LOGICAL_AGGR).addChildren(Pattern.create(
+                        OperatorType.LOGICAL_OLAP_SCAN)));
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        OptExpression scanExpression = input.getInputs().get(0);
+        Preconditions.checkState(scanExpression.getOp() instanceof LogicalOlapScanOperator);
+        LogicalOlapScanOperator scanOperator = (LogicalOlapScanOperator) scanExpression.getOp();
+        if (scanOperator.getProjection() != null) {
+            return false;
+        }
+        OlapTable olapTable = (OlapTable) scanOperator.getTable();
+
+        MaterializedIndexMeta materializedIndexMeta = olapTable.getIndexMetaByIndexId(scanOperator.getSelectedIndexId());
+        // must be Aggregation
+        if (!materializedIndexMeta.getKeysType().isAggregationFamily()) {
+            return false;
+        }
+        Set<String> keyColumnNames = Sets.newHashSet();
+        for (Column column : materializedIndexMeta.getSchema()) {
+            if (column.isKey()) {
+                keyColumnNames.add(column.getName().toLowerCase());
+            }
+        }
+
+        // group by keys contain partition columns and distribution columns
+        PartitionInfo partitionInfo = olapTable.getPartitionInfo();
+        List<String> partitionColumnNames = Lists.newArrayList();
+        if (partitionInfo instanceof RangePartitionInfo) {
+            RangePartitionInfo rangePartitionInfo = (RangePartitionInfo) partitionInfo;
+            partitionColumnNames.addAll(rangePartitionInfo.getPartitionColumns().stream()
+                    .map(column -> column.getName().toLowerCase()).collect(Collectors.toList()));
+        }
+
+        List<String> distributionColumnNames = olapTable.getDistributionColumnNames().stream()
+                .map(String::toLowerCase).collect(Collectors.toList());
+
+        LogicalAggregationOperator aggregationOperator = (LogicalAggregationOperator) input.getOp();
+        for (Map.Entry<ColumnRefOperator, CallOperator> entry : aggregationOperator.getAggregations().entrySet()) {
+            if (unsupportedFunctionNames.contains(entry.getValue().getFnName().toLowerCase())) {
```

Comment:
need check the input of aggregate function is materialized column?

Reply:
done